src/dsp/interface.ts

```diff
@@ -1,56 +1,40 @@
-/* eslint-disable @typescript-eslint/no-unused-vars */
 // - prepare data from yaml
 // - fill input
 // - prep structure
 
+import type { KeyofToArray } from '@murongg/utils'
+import type { Input } from '../common/input'
+import { dspParamsNormalize } from '../common/utils'
 import type { ZkGraphYaml } from '../types/zkgyaml'
 
 export class DataPrep {}
 
-export class DataSourcePlugin {
-  static getLibDSPName() {
-    // SHOULD align with zkgraph-lib/dsp/ethereum/index.ts
-    // return ['zkmain_name_in_lib', 'asmain_name_in_lib']
-    throw new Error('default: getLibDSPName not implemented in DSP.')
-  }
-
-  static async prepareData(zkgraphYaml: ZkGraphYaml, prepareParams: Record<string, any>): Promise<any> {
-    throw new Error('default: prepareData not implemented in DSP.')
-  }
-
-  static fillExecInput(input: any, zkgraphYaml: any, dataPrep: any) {
-    throw new Error('default: fillInput not implemented in DSP.')
-  }
-
-  static fillProveInput(input: any, zkgraphYaml: any, dataPrep: any) {
-    throw new Error('default: fillInput not implemented in DSP.')
-  }
-
-  static toProveDataPrep(execDataPrep: any, execResult: any) {
-    throw new Error('default: toProveDataPrep not implemented in DSP.')
-  }
+export abstract class DataSourcePlugin<EP extends object, PP extends object, PRP extends object, DPP extends object> {
```

Comment:
good to have EP, PP, PRP.
but do we need DPP? 1. it's only DP class not DP Params, so DP is better than DPP.
2. what's the benefit to have DP in the construction generics? im thinking of keep DP free and out of generics.

Reply:
When developing DSP, it is possible to omit the return type after the corresponding method.
e.g.
- no `DPP`
```ts
toProveDataPrep(...args): DataPrep
```
- use `DPP`
```ts
toProveDataPrep(...args)
```
You can omit the return type and direct use inheritance of the `DataSourcePlugin` class.